src/docs/05-go-client/19-workflow-non-deterministic-error.md

```diff
@@ -0,0 +1,178 @@
+```go
+workflow.Sleep(time.Hour)
+```
+and restart worker, then it will run into this error. Because in the history, the workflow has a timer event that is supposed to fire in one hour. However, during replay, there is no logic to schedule that timer.
+
+### 2. Extra decisions
+```go
+fmt.Errorf("nondeterministic workflow: extra replay decision for %s", util.DecisionToString(d))
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_task_handlers.go#L1210)
+
+This is basically the opposite of the previous case, which means that during replay, Cadence generates more decisions than those in history events. Using the previous history as an example, when the workflow is waiting at the one hour timer(event ID 22), if we change the line of:
+```go
+err = workflow.ExecuteActivity(ctx, activityB, a).Get(ctx, nil)
+```
+to
+```go
+fb := workflow.ExecuteActivity(ctx, activityB, a)
+fc := workflow.ExecuteActivity(ctx, activityC, a)
+err = fb.Get(ctx,nil)
+if err != nil {
+	return err
+}
+err = fc.Get(ctx,nil)
+if err != nil {
+	return err
+}
+```
+And restart worker, then it will run into this error. Because in the history, the workflow has scheduled only activityB after the one minute timer, however, during replay, there are two activities scheduled in a decision( in parallel).
+
+### 3. Mismatched decisions
+```go
+fmt.Errorf("nondeterministic workflow: history event is %s, replay decision is %s",util.HistoryEventToString(e), util.DecisionToString(d))
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_task_handlers.go#L1214)
+
+This means after replay code, the decision scheduled is different than the one in history. Using the previous history as an example, when the workflow is waiting at the one hour timer(event ID 22),
+if we change the line of :
+```go
+err = workflow.ExecuteActivity(ctx, ActivityB, a).Get(ctx, nil)
+```
+to
+```go
+err = workflow.ExecuteActivity(ctx, ActivityC, a).Get(ctx, nil)
+```
+And restart worker, then it will run into this error. Because in the history, the workflow has scheduled ActivityB with input a, but during replay, it schedules ActivityC.
+
+### 4. Decision state machine panic
+```go
+fmt.Sprintf("unknown decision %v, possible causes are nondeterministic workflow definition code"+" or incompatible change in the workflow definition", id)
+```
+For source code click [here](https://github.com/uber-go/cadence-client/blob/e5081b085b0333bac23f198e57959681e0aee987/internal/internal_decision_state_machine.go#L693)
+
+This usually means workflow history is corrupted due to some bug. For example, the same activity can be scheduled and differentiated by activityID. So ActivityIDs for different activities are supposed to be unique in workflow history. If however we have an ActivityID collision, replay will run into this error.
```

Comment:
we should probably check this more closely, but it might be true.  I think it's currently thrown for other scenarios though, including "it replayed blindly too far before validating, and did something irrational instead of detecting the non-determinism earlier"